### PR TITLE
Defining more consistent API across ArrayList and DataList

### PR DESCRIPTION
**Pull request by dhensby**
see #3828

### CODE REVIEW

File: model/ArrayList.php [hunk 1]
```diff
@@ -477,6 +477,72 @@ public function canFilterBy($by) {
...
+			}
+		}
+
+		$list = clone $this;
+		$list->items = $itemsToKeep;
+		return $list;
+
+	}
+
+	protected function normaliseFilterArgs() {
```
Comment (tractorcow) [inline model/ArrayList.php]: PHPDoc please. :D

File: model/ArrayList.php [hunk 2]
```diff
@@ -477,6 +477,72 @@ public function canFilterBy($by) {
...
+	 * @return DataList
+	 */
+	public function filterAny() {
+		$keepUs = call_user_func_array(array($this, 'normaliseFilterArgs'), func_get_args());
+
+		$itemsToKeep = array();
+
+		foreach ($this->items as $column => $value) {
+			$keepItem = false;
+			if ((is_array($value) && in_array($this->extractValue($item, $column), $value))
```
Comment (camspiers) [inline model/ArrayList.php]: It seems like there is some code missing here. Don't you want to iterate over $keepUs and for any passing conditions set keepItem true then break?

Probably a good idea to add tests for this too

Comment (tractorcow) [inline model/ArrayList.php]: Yep, just look at the deleted code in `filter` for this logic.

**Comment by tractorcow** [comment]
I've also added some comments to https://github.com/silverstripe/silverstripe-framework/pull/3828.

**Comment by dhensby** [comment]
feedback from other PR incorporated here

**Comment by dhensby** [comment]
bump @tractorcow

**Comment by tractorcow** [comment]
Rebased and merged with https://github.com/silverstripe/silverstripe-framework/commit/53ffc1a0dbac52ad682a150d387b104c1dc0616f

**Pull request closed without merging**